Gems/Terrain/Code/Source/TerrainRenderer/TerrainClipmapManager.cpp

```diff
@@ -0,0 +1,249 @@
+        UpdateClipmapData(cameraPosition);
+        terrainSrg->SetConstant(m_clipmapDataIndex, m_clipmapData);
+
+        if (m_clipmapsNeedUpdate)
+        {
+            terrainSrg->SetImage(m_macroColorClipmapsIndex, m_clipmaps.at(AZ::Name(MacroColorClipmapString)));
+            terrainSrg->SetImage(m_macroNormalClipmapsIndex, m_clipmaps.at(AZ::Name(MacroNormalClipmapString)));
+            terrainSrg->SetImage(m_detailColorClipmapsIndex, m_clipmaps.at(AZ::Name(DetailColorClipmapString)));
+            terrainSrg->SetImage(m_detailNormalClipmapsIndex, m_clipmaps.at(AZ::Name(DetailNormalClipmapString)));
+            terrainSrg->SetImage(m_detailHeightClipmapsIndex, m_clipmaps.at(AZ::Name(DetailHeightClipmapString)));
+            terrainSrg->SetImage(m_detailMiscClipmapsIndex, m_clipmaps.at(AZ::Name(DetailMiscClipmapString)));
+
+            m_clipmapsNeedUpdate = false;
+        }
+    }
+
+    void TerrainClipmapManager::InitializeClipmapData()
+    {
+        float clipmapScaleInv = 1.0f;
+        for (int32_t clipmapIndex = ClipmapStackSize - 1; clipmapIndex >= 0; --clipmapIndex)
+        {
+            AZ::Vector4(0.5f).StoreToFloat4(m_clipmapData.m_clipmapCenters[clipmapIndex].data());
+            AZ::Vector4(clipmapScaleInv).StoreToFloat4(m_clipmapData.m_clipmapScaleInv[clipmapIndex].data());
+            clipmapScaleInv /= 2.0f;
+        }
+
+        AZ::Vector4::CreateZero().StoreToFloat4(m_clipmapData.m_viewPosition.data());
+
+        m_clipmapData.m_clipmapSize[0] = ClipmapSizeWidth;
+        m_clipmapData.m_clipmapSize[1] = ClipmapSizeHeight;
+
+        AZ::Aabb worldBounds = AZ::Aabb::CreateNull();
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            worldBounds, &AzFramework::Terrain::TerrainDataRequests::GetTerrainAabb);
+
+        m_clipmapData.m_worldBounds[0] = worldBounds.GetMin().GetX();
+        m_clipmapData.m_worldBounds[1] = worldBounds.GetMin().GetY();
+        m_clipmapData.m_worldBounds[2] = worldBounds.GetMax().GetX();
+        m_clipmapData.m_worldBounds[3] = worldBounds.GetMax().GetY();
+
+        // Use world size for now.
+        const AZ::Vector3 worldSize = worldBounds.GetMax() - worldBounds.GetMin();
+        AZ::Vector2(worldSize.GetX(), worldSize.GetY()).StoreToFloat2(m_clipmapData.m_maxRenderSize.data());
+    }
+
+    void TerrainClipmapManager::InitializeClipmapImages()
+    {
+        AZ::Data::Instance<AZ::RPI::AttachmentImagePool> pool = AZ::RPI::ImageSystemInterface::Get()->GetSystemAttachmentPool();
+
+        AZ::RHI::ImageDescriptor imageDesc;
+        imageDesc.m_format = AZ::RHI::Format::R8G8B8A8_UNORM;
+        imageDesc.m_bindFlags = AZ::RHI::ImageBindFlags::ShaderReadWrite;
+        imageDesc.m_size = AZ::RHI::Size(ClipmapSizeWidth, ClipmapSizeHeight, 1);
+        imageDesc.m_arraySize = ClipmapStackSize;
+
+        AZ::Name macroColorClipmapName = AZ::Name(MacroColorClipmapString);
+        AZ::Data::Instance<AZ::RPI::AttachmentImage> macroColorClipmaps =
```

Comment:
Should we use a format with a gamma transfer for the color data? Or is the issue that you can't use `R8G8B8A8_SRGB` since you're writing directly from a compute shader?

Reply:
Same as normal map. Because I am reusing GetDetailNormal/GetDetailColor for clipmaps. I'll change and do the color space changing later.